File: src/devkit/platform.py
"""Host OS detection for Windows, macOS, and Linux."""

from __future__ import annotations

import os
import platform
from enum import Enum
from pathlib import Path


class HostOS(str, Enum):
    """Supported operating systems."""

    WINDOWS = "windows"
    MACOS = "macos"
    LINUX = "linux"
    OTHER = "other"


def current_os() -> HostOS:
    """Return the current host OS."""
    system = platform.system()
    if system == "Windows":
        return HostOS.WINDOWS
    if system == "Darwin":
        return HostOS.MACOS
    if system == "Linux":
        return HostOS.LINUX
    return HostOS.OTHER


def is_windows() -> bool:
    return current_os() == HostOS.WINDOWS


def is_macos() -> bool:
    return current_os() == HostOS.MACOS


def is_linux() -> bool:
    return current_os() == HostOS.LINUX


def is_unix() -> bool:
    return current_os() in {HostOS.MACOS, HostOS.LINUX, HostOS.OTHER}


def cpu_arch() -> str:
    """Return a normalized CPU arch label: ``x64`` or ``aarch64``."""
    machine = platform.machine().lower()
    if machine in {"arm64", "aarch64"}:
        return "aarch64"
    if machine in {"x86_64", "amd64", "x64"}:
        return "x64"
    if machine in {"i386", "i686", "x86"}:
        return "x86"
    return machine


def adoptium_os() -> str:
    """OS slug for the Adoptium / Temurin API."""
    host = current_os()
    if host == HostOS.WINDOWS:
        return "windows"
    if host == HostOS.MACOS:
        return "mac"
    if host == HostOS.LINUX:
        return "linux"
    raise RuntimeError(f"Unsupported OS for Temurin JDK: {host.value}")


def os_label() -> str:
    """Human-readable OS name for CLI output."""
    labels = {
        HostOS.WINDOWS: "Windows",
        HostOS.MACOS: "macOS",
        HostOS.LINUX: "Linux",
        HostOS.OTHER: platform.system() or "Unknown",
    }
    return labels[current_os()]


def pick_for_os(mapping: dict[str, str], *, default: str | None = None) -> str:
    """Pick a value from a platform map.

    Keys may be ``windows``, ``macos``/``darwin``, ``linux``, or ``*``/``default``.
    """
    os_key = current_os().value
    aliases = {
        HostOS.MACOS: ("macos", "darwin", "osx"),
        HostOS.WINDOWS: ("windows", "win", "win32"),
        HostOS.LINUX: ("linux",),
        HostOS.OTHER: ("other",),
    }
    lowered = {k.lower(): v for k, v in mapping.items()}
    for alias in aliases.get(current_os(), (os_key,)):
        if alias in lowered:
            return lowered[alias]
    if "default" in lowered:
        return lowered["default"]
    if "*" in lowered:
        return lowered["*"]
    if default is not None:
        return default
    known = ", ".join(sorted(mapping))
    raise KeyError(f"No URL/value for OS '{os_key}'. Available keys: {known}")


def shell_profile_candidates() -> list[Path]:
    """Return likely shell profile paths for the current OS (macOS/Linux)."""
    home = Path.home()
    shell = os.environ.get("SHELL", "")
    profiles: list[Path] = []

    if "zsh" in shell or current_os() == HostOS.MACOS:
        profiles.extend([home / ".zshrc", home / ".zprofile"])
    if "bash" in shell or current_os() == HostOS.LINUX:
        profiles.extend([home / ".bashrc", home / ".bash_profile", home / ".profile"])
    # Always include common fallbacks without duplicates
    for extra in (home / ".zshrc", home / ".bashrc", home / ".profile"):
        if extra not in profiles:
            profiles.append(extra)
    return profiles


def primary_shell_profile() -> Path:
    """Best profile file to source ``~/.devkit/env.sh`` from."""
    shell = os.environ.get("SHELL", "")
    home = Path.home()
    if "zsh" in shell or current_os() == HostOS.MACOS:
        return home / ".zshrc"
    if "bash" in shell:
        # Prefer .bashrc when it exists (typical Linux interactive shells)
        bashrc = home / ".bashrc"
        if bashrc.exists() or current_os() == HostOS.LINUX:
            return bashrc
        return home / ".bash_profile"
    return home / ".profile"

File: src/platform.rs
//! Host OS detection for Windows, macOS, and Linux.

use std::path::PathBuf;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HostOS {
    Windows,
    MacOS,
    Linux,
    Other,
}

pub fn current_os() -> HostOS {
    if cfg!(target_os = "windows") {
        HostOS::Windows
    } else if cfg!(target_os = "macos") {
        HostOS::MacOS
    } else if cfg!(target_os = "linux") {
        HostOS::Linux
    } else {
        HostOS::Other
    }
}

pub fn is_windows() -> bool {
    current_os() == HostOS::Windows
}

pub fn is_macos() -> bool {
    current_os() == HostOS::MacOS
}

pub fn is_linux() -> bool {
    current_os() == HostOS::Linux
}

pub fn is_unix() -> bool {
    matches!(current_os(), HostOS::MacOS | HostOS::Linux | HostOS::Other)
}

/// Return a normalized CPU arch label: `x64` or `aarch64`.
pub fn cpu_arch() -> String {
    let machine = std::env::consts::ARCH.to_lowercase();
    match machine.as_str() {
        "aarch64" | "arm64" => "aarch64".to_string(),
        "x86_64" | "amd64" | "x64" => "x64".to_string(),
        "x86" | "i386" | "i686" => "x86".to_string(),
        other => other.to_string(),
    }
}

/// OS slug for the Adoptium / Temurin API.
pub fn adoptium_os() -> anyhow::Result<&'static str> {
    Ok(match current_os() {
        HostOS::Windows => "windows",
        HostOS::MacOS => "mac",
        HostOS::Linux => "linux",
        HostOS::Other => anyhow::bail!("Unsupported OS for Temurin JDK: other"),
    })
}

/// Human-readable OS name for CLI output.
pub fn os_label() -> &'static str {
    match current_os() {
        HostOS::Windows => "Windows",
        HostOS::MacOS => "macOS",
        HostOS::Linux => "Linux",
        HostOS::Other => "Unknown",
    }
}

/// Pick a value from a platform map.
///
/// Keys may be `windows`, `macos`/`darwin`, `linux`, or `*`/`default`.
pub fn pick_for_os(mapping: &[(&str, &str)], default: Option<&str>) -> anyhow::Result<String> {
    let aliases: &[&str] = match current_os() {
        HostOS::MacOS => &["macos", "darwin", "osx"],
        HostOS::Windows => &["windows", "win", "win32"],
        HostOS::Linux => &["linux"],
        HostOS::Other => &["other"],
    };
    let lowered: Vec<(String, &str)> = mapping
        .iter()
        .map(|(k, v)| (k.to_lowercase(), *v))
        .collect();
    for alias in aliases {
        if let Some((_, v)) = lowered.iter().find(|(k, _)| k == alias) {
            return Ok(v.to_string());
        }
    }
    if let Some((_, v)) = lowered.iter().find(|(k, _)| k == "default") {
        return Ok(v.to_string());
    }
    if let Some((_, v)) = lowered.iter().find(|(k, _)| k == "*") {
        return Ok(v.to_string());
    }
    if let Some(d) = default {
        return Ok(d.to_string());
    }
    let known: Vec<&str> = mapping.iter().map(|(k, _)| *k).collect();
    anyhow::bail!(
        "No URL/value for OS '{}'. Available keys: {}",
        std::env::consts::OS,
        known.join(", ")
    )
}

/// Return likely shell profile paths for the current OS (macOS/Linux).
pub fn shell_profile_candidates() -> Vec<PathBuf> {
    let home = dirs::home_dir().unwrap_or_default();
    let shell = std::env::var("SHELL").unwrap_or_default();
    let mut profiles: Vec<PathBuf> = Vec::new();

    if shell.contains("zsh") || current_os() == HostOS::MacOS {
        profiles.push(home.join(".zshrc"));
        profiles.push(home.join(".zprofile"));
    }
    if shell.contains("bash") || current_os() == HostOS::Linux {
        profiles.push(home.join(".bashrc"));
        profiles.push(home.join(".bash_profile"));
        profiles.push(home.join(".profile"));
    }
    for extra in [
        home.join(".zshrc"),
        home.join(".bashrc"),
        home.join(".profile"),
    ] {
        if !profiles.contains(&extra) {
            profiles.push(extra);
        }
    }
    profiles
}

/// Best profile file to source `~/.devkit/env.sh` from.
pub fn primary_shell_profile() -> PathBuf {
    let home = dirs::home_dir().unwrap_or_default();
    let shell = std::env::var("SHELL").unwrap_or_default();

    if shell.contains("zsh") || current_os() == HostOS::MacOS {
        return home.join(".zshrc");
    }
    if shell.contains("bash") {
        let bashrc = home.join(".bashrc");
        if bashrc.exists() || current_os() == HostOS::Linux {
            return bashrc;
        }
        return home.join(".bash_profile");
    }
    home.join(".profile")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn current_os_matches_target() {
        let os = current_os();
        assert!(os == HostOS::Windows || is_unix());
    }

    #[test]
    fn pick_for_os_falls_back_to_default_key() {
        let v = pick_for_os(&[("default", "everywhere")], None).unwrap();
        assert_eq!(v, "everywhere");
    }

    #[test]
    fn pick_for_os_errors_without_match() {
        assert!(pick_for_os(&[("bogus", "x")], None).is_err());
    }

    #[test]
    fn cpu_arch_is_known_label() {
        let arch = cpu_arch();
        assert!(["x64", "aarch64", "x86"].contains(&arch.as_str()) || !arch.is_empty());
    }
}
